source/http_message.c

```diff
@@ -238,15 +224,16 @@ static napi_value s_request_headers_get(napi_env env, void *native_this) {
 
     napi_value result = NULL;
     if (binding->node_headers) {
+        /* This HTTP request already has a reference to the Node object for the headers */
         AWS_NAPI_ENSURE(env, napi_get_reference_value(env, binding->node_headers, &result));
     } else {
+        /* There is no Node object for these headers.
+         * Create a Node object, and a reference to it, and store that reference on this HTTP request */
         struct aws_http_headers *headers = aws_http_message_get_headers(binding->native);
         AWS_NAPI_ENSURE(env, aws_napi_http_headers_wrap(env, headers, &result));
+        AWS_NAPI_ENSURE(env, napi_create_reference(env, result, 1, &binding->node_headers));
     }
 
-    /* Store the value for later */
-    AWS_NAPI_ENSURE(env, napi_create_reference(env, result, 1, &binding->node_headers));
-
     return result;
```

Comment:
This was the leak. We were creating a new `napi_ref` each time the headers were queried. Now, we only create a `napi_ref` if we didn't already have one.